proposals/nnnn-atf-selenium-support.md

```diff
@@ -0,0 +1,236 @@
+
+#### HMI Websocket Server
+
+ATF will need to create a websocket server that will receive messages from the HMI. This server will then be routed through the ATF test logic to verify the contents of the HMI messages. After verifying that the messages are valid, ATF will pass the HMI message to SDL Core.
+
+#### Core Websocket Client
+
+ATF already has a websocket client that connects to SDL Core but this piece will need to be extended to allow for validation of messages before they are passed to the HMI.
+
+### ATF Test Logic
+
+When ATF is configured to use selenium, there are a few methods used in ATF that should be extended.
+
+This is an existing simple test for a Show RPC request.
+
+```lua
+local function show(pParams)
+  local cid = common.getMobileSession():SendRPC("Show", pParams.requestParams)
+  pParams.responseUiParams.appID = common.getHMIAppId()
+  common.getHMIConnection():ExpectRequest("UI.Show", pParams.responseUiParams)
+  :Do(function(_, data)
+      common.getHMIConnection():SendResponse(data.id, data.method, "SUCCESS", { })
+    end)
+  common.getMobileSession():ExpectResponse(cid, { success = true, resultCode = "SUCCESS" })
+end
+```
+
+#### Extend ExpectRequest
+
+In this code example there is a simulated mobile request being passed to SDL Core. The simulated HMI has a defined behavior for handling the request via ExpectRequest. For selenium compatible tests, ATF's ExpectRequest() should be extended to forward the RPC to the HMI and then the selenium web driver can verify the state of the HMI if applicable.
+
+#### Extend SendResponse
+
+After handling the request, the HMI will automatically respond with a success response. ATF will need to catch this response from the HMI and match it with the parameters included in the test. If the parameters match, ATF will forward the message to SDL core and the test will continue. To implement this behavior, the SendResponse() method will need to be extended for selenium compatible test scripts.
+
+#### Extend SendNotification
+
+Most notifications sent by the HMI are triggered as a result of a user interaction a with the HMI. Examples of these events are pressing a menu command, or consenting to app permissions.
+
+```lua
+local function onCommand()
+  common.getHMIConnection():SendNotification("UI.OnCommand", { cmdID = 20, appID = common.getHMIAppId() })
+  common.getMobileSession():ExpectNotification("OnCommand", { cmdID = 20, triggerSource= "MENU" })
+end
+```
+
+SendNotification should be extended to notify the selenium web driver it should simulate a user action if applicable. In this case it would be selecting menu command with cmdID = 20. This will trigger the HMI to send the `UI.OnCommand` RPC. ATF will need to listen for this notification and match it with the parameters listed in the test. ATF will then forward the notification from the HMI to SDL Core after validating the msg contents.
```

Comment:
```suggestion
`SendNotification` should be extended to notify the Selenium WebDriver that it should simulate a user action if applicable. In this case it would be selecting menu command with `cmdID = 20`. This will trigger the HMI to send the `UI.OnCommand` RPC. ATF will need to listen for this notification and match it with the parameters listed in the test. ATF will then forward the notification from the HMI to SDL Core after validating the message contents.
```